modules/currencies/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+			let source = T::AddressMapping::to_evm_address(&from).ok_or(Error::<T>::AccountNotFound)?;
+			let address = T::AddressMapping::to_evm_address(&to).ok_or(Error::<T>::AccountNotFound)?;
+			T::EVMBridge::transfer(InvokeContext { contract, source }, address, amount)?;
+		} else if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::transfer(from, to, amount)?;
+		} else {
+			T::MultiCurrency::transfer(currency_id, from, to, amount)?;
+		}
+
+		Self::deposit_event(RawEvent::Transferred(currency_id, from.clone(), to.clone(), amount));
+		Ok(())
+	}
+
+	fn deposit(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		Self::ensure_not_erc20(currency_id)?;
+		if amount.is_zero() {
+			return Ok(());
+		}
+		if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::deposit(who, amount)?;
+		} else {
+			T::MultiCurrency::deposit(currency_id, who, amount)?;
+		}
+		Self::deposit_event(RawEvent::Deposited(currency_id, who.clone(), amount));
+		Ok(())
+	}
+
+	fn withdraw(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		Self::ensure_not_erc20(currency_id)?;
+		if amount.is_zero() {
+			return Ok(());
+		}
+		if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::withdraw(who, amount)?;
+		} else {
+			T::MultiCurrency::withdraw(currency_id, who, amount)?;
+		}
+		Self::deposit_event(RawEvent::Withdrawn(currency_id, who.clone(), amount));
+		Ok(())
+	}
+
+	fn can_slash(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> bool {
+		if currency_id.is_erc20() {
+			false
+		} else if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::can_slash(who, amount)
+		} else {
+			T::MultiCurrency::can_slash(currency_id, who, amount)
+		}
+	}
+
+	fn slash(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> Self::Balance {
+		if currency_id.is_erc20() {
+			Default::default()
+		} else if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::slash(who, amount)
+		} else {
+			T::MultiCurrency::slash(currency_id, who, amount)
+		}
+	}
+}
+
+impl<T: Config> MultiCurrencyExtended<T::AccountId> for Module<T> {
+	type Amount = AmountOf<T>;
+
+	fn update_balance(currency_id: Self::CurrencyId, who: &T::AccountId, by_amount: Self::Amount) -> DispatchResult {
+		Self::ensure_not_erc20(currency_id)?;
+		if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::update_balance(who, by_amount)?;
+		} else {
+			T::MultiCurrency::update_balance(currency_id, who, by_amount)?;
+		}
+		Self::deposit_event(RawEvent::BalanceUpdated(currency_id, who.clone(), by_amount));
+		Ok(())
+	}
+}
+
+impl<T: Config> MultiLockableCurrency<T::AccountId> for Module<T> {
+	type Moment = T::BlockNumber;
+
+	fn set_lock(lock_id: LockIdentifier, currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) {
```

Comment:
I think we will need to modify `MultiLockableCurrency` to have those return `DispatchResult` so we don't ignore error here

Reply:
return error when currency id is ERC20?

Reply:
yes. that's better than ignore error.